File: scripts/test-rpg-assets.mjs
// Real WASM + native replay: changing a PNG requires no rebuild.
import assert from 'node:assert/strict';
import { createRequire } from 'node:module';
import { execFile, run } from './acceptance_process.mjs';
import { readFileSync, writeFileSync, mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { resolve, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { deflateSync } from 'node:zlib';
const repo = fileURLToPath(new URL('../', import.meta.url));
const metadata = JSON.parse(await execFile('cargo', ['metadata', '--no-deps', '--format-version', '1'], { phase: 'build', cwd: repo, encoding: 'utf8' }));
const require = createRequire(import.meta.url);
const { BrowserRuntime, BrowserLiveRuntime, verify_recording_json_with_pngs } = require(resolve(metadata.target_directory, 'titan/browser-node/titan_browser.js'));
const original = readFileSync(join(repo, 'assets/player.png'));
const originalTree = readFileSync(join(repo, 'assets/tree.png'));
function crc32(bytes) {
  let crc = 0xffffffff;
  for (const byte of bytes) { crc ^= byte; for (let i = 0; i < 8; i++) crc = (crc >>> 1) ^ ((crc & 1) ? 0xedb88320 : 0); }
  return (crc ^ 0xffffffff) >>> 0;
}
function chunk(type, data) {
  const body = Buffer.concat([Buffer.from(type), data]);
  const head = Buffer.alloc(4), tail = Buffer.alloc(4);
  head.writeUInt32BE(data.length); tail.writeUInt32BE(crc32(body));
  return Buffer.concat([head, body, tail]);
}
function png(width, height) {
  const header = Buffer.alloc(13);
  header.writeUInt32BE(width); header.writeUInt32BE(height, 4); header[8] = 8; header[9] = 6;
  const rows = Buffer.alloc(height * (1 + width * 4));
  for (let y = 0; y < height; y++) for (let x = 0; x < width; x++) rows.set([240, 30, 180, 255], y * (1 + width * 4) + 1 + x * 4);
  return Buffer.concat([Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]), chunk('IHDR', header), chunk('IDAT', deflateSync(rows)), chunk('IEND', Buffer.alloc(0))]);
}
const replacement = png(8, 8);
let sequence = 0;
function call(runtime, request, success = true) {
  const result = JSON.parse(runtime.handle(JSON.stringify({ schema_version: 1, request_id: `asset-${++sequence}`, request })));
  assert.equal(result.status, success ? 'success' : 'failure', JSON.stringify(result));
  return result;
}
const query = (runtime, name) => call(runtime, { type: 'query', name, arguments: {} }).response.value;
const capture = runtime => call(runtime, { type: 'capture' }).response.checksum;
const invoke = (runtime, name, args = {}) => call(runtime, { type: 'invoke', name, arguments: args });
function reference(runtime) {
  let frame = call(runtime, { type: 'status' }).observed_frame;
  for (const [action, ticks] of [['right', 2], ['down', 3], ['right', 6]]) for (let tick = 0; tick < ticks; tick++) {
    call(runtime, { type: 'inject_input', frame: ++frame, actions: { [action]: { kind: 'button', value: true } } });
  }
  call(runtime, { type: 'step', frames: 11 });
}
for (const factory of [(player, tree) => BrowserRuntime.with_pngs(true, player, tree), (player, tree) => BrowserLiveRuntime.with_pngs(player, tree)]) {
  for (const bytes of [new Uint8Array(), original.subarray(0, 20), Buffer.from('not a PNG'), new Uint8Array(256 * 1024 + 1), png(65, 1)]) {
    assert.throws(() => factory(bytes, originalTree), /player.png/);
    assert.throws(() => factory(original, bytes), /tree.png/);
  }
  factory(original, originalTree).free(); // Repaired construction after either source failed.
}
const procedural = new BrowserRuntime(true);
const loaded = BrowserRuntime.with_pngs(true, original, originalTree);
assert.equal(capture(loaded), capture(procedural));
reference(loaded); reference(procedural);
assert.equal(capture(loaded), 'f7a298f62ad75c1c');
assert.deepEqual(query(loaded, 'save'), query(procedural, 'save'));

await execFile('cargo', ['build', '--example', 'replay_rpg'], { phase: 'build', cwd: repo, stdio: 'inherit' });
const temporary = mkdtempSync(join(tmpdir(), 'titan-rpg-png-'));
try {
  const checksums = new Set([capture(procedural)]);
  for (const [playerBytes, treeBytes] of [[replacement, originalTree], [original, replacement], [replacement, replacement]]) {
    const changed = BrowserRuntime.with_pngs(true, playerBytes, treeBytes);
    reference(changed);
    assert.deepEqual(query(changed, 'save'), query(procedural, 'save'));
    const changedReference = capture(changed);
    assert.ok(!checksums.has(changedReference), 'each independently replaced sprite changes pixels');
    checksums.add(changedReference); changed.free();

    const live = BrowserLiveRuntime.with_pngs(playerBytes, treeBytes);
    live.set_control_enabled(true);
    const initialChecksum = capture(live);
    reference(live); assert.equal(capture(live), changedReference);
    const save = query(live, 'save');
    const recording = JSON.stringify(query(live, 'recording'));
    assert.equal(JSON.parse(verify_recording_json_with_pngs(recording, playerBytes, treeBytes)).verified, true);
    if (playerBytes === replacement) assert.throws(() => verify_recording_json_with_pngs(recording, original, treeBytes), /mismatch|asset/i);
    if (treeBytes === replacement) assert.throws(() => verify_recording_json_with_pngs(recording, playerBytes, originalTree), /mismatch|asset/i);
    // All reconstruction paths retain the complete startup pair.
    invoke(live, 'restart'); assert.equal(capture(live), initialChecksum);
    invoke(live, 'load_save', { save }); assert.equal(capture(live), changedReference);
    invoke(live, 'restart');
    live.load_recording(recording);
    while (!JSON.parse(live.playback_status()).complete) live.step_playback();
    assert.equal(JSON.parse(live.playback_status()).verified, true);
    assert.equal(capture(live), changedReference);
    live.restart_playback(); assert.equal(capture(live), initialChecksum);
    live.exit_playback(); assert.equal(capture(live), initialChecksum);
    reference(live); assert.equal(capture(live), changedReference); live.free();

    writeFileSync(join(temporary, 'player.png'), playerBytes);
    writeFileSync(join(temporary, 'tree.png'), treeBytes);
    const path = join(temporary, 'recording.json'); writeFileSync(path, recording);
    const executable = resolve(metadata.target_directory, 'debug/examples/replay_rpg');
    const native = JSON.parse(await execFile(executable, [path, '--assets-dir', temporary], { cwd: temporary, encoding: 'utf8' }));
    assert.equal(native.verified, true); assert.equal(native.checksum, changedReference);
    const wrong = await run(executable, [path, '--assets-dir', join(repo, 'assets')], { cwd: temporary, encoding: 'utf8' });
    assert.notEqual(wrong.status, 0); assert.match(wrong.stderr, /mismatch|asset/i);
    for (const [name, bytes] of [['player.png', playerBytes], ['tree.png', treeBytes]]) {
      rmSync(join(temporary, name));
      const missing = await run(executable, [path, '--assets-dir', temporary], { cwd: temporary, encoding: 'utf8' });
      assert.notEqual(missing.status, 0); assert.ok(missing.stderr.includes(name));
      writeFileSync(join(temporary, name), bytes);
    }
  }
} finally { rmSync(temporary, { recursive: true, force: true }); }
loaded.free(); procedural.free();
console.log('Two-sprite WASM/native acceptance passed: reference pixels, independent replacements, bounded failures, retained pair and replay verification.');

File: scripts/test-rpg-assets.py
#!/usr/bin/env python3
"""Bounded file-backed RPG checks; --gpu additionally checks a relocated app bundle."""
import argparse
import json
import os
from pathlib import Path
import re
import shutil
import struct
import acceptance_process as processes
import sys
import tempfile
import time
import zlib

ROOT = Path(__file__).resolve().parents[1]
REFERENCE = 'f7a298f62ad75c1c'


def png(color, width=8, height=10):
    def chunk(kind, data):
        return struct.pack('>I', len(data)) + kind + data + struct.pack('>I', zlib.crc32(kind + data))
    pixels = (b'\0' + bytes(color) * width) * height
    return (b'\x89PNG\r\n\x1a\n' + chunk(b'IHDR', struct.pack('>IIBBBBB', width, height, 8, 6, 0, 0, 0))
            + chunk(b'IDAT', zlib.compress(pixels)) + chunk(b'IEND', b''))


def run(command, cwd, success=True, phase="runtime"):
    result = processes.run(list(map(str, command)), cwd=cwd, capture_output=True, text=True, phase=phase)
    assert (result.returncode == 0) == success, (command, result.stdout, result.stderr)
    return result


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--gpu', action='store_true')
    options = parser.parse_args()
    build = ['cargo', 'build', '--locked', '-p', 'titan-cli', '-p', 'titan', '--bin', 'titan',
             '--example', 'procedural_rpg', '--example', 'replay_rpg']
    run(build, ROOT, phase="build")
    metadata = json.loads(run(['cargo', 'metadata', '--locked', '--no-deps', '--format-version', '1'], ROOT, phase='build').stdout)
    target = Path(metadata['target_directory'])
    executable = target / 'debug/examples/procedural_rpg'
    verifier = target / 'debug/examples/replay_rpg'
    cli = target / 'debug/titan'
    with tempfile.TemporaryDirectory(prefix='titan asset acceptance ') as directory:
        project = Path(directory)
        assets = project / 'art outside checkout'
        assets.mkdir()
        sprite = assets / 'player.png'
        sprite.write_bytes((ROOT / 'assets/player.png').read_bytes())
        tree = assets / 'tree.png'
        tree.write_bytes((ROOT / 'assets/tree.png').read_bytes())

        def reference(*arguments):
            output = run([executable, *arguments], project).stdout
            match = re.search(r'\(3 shards, shrine active: true, checksum: ([0-9a-f]{16})\)', output)
            assert match, output
            return match[1]

        assert reference('--generated-assets') == REFERENCE
        assert reference('--assets-dir', assets) == REFERENCE
        # Reuse the same executable after replacing an external asset; no Cargo call occurs here.
        binary_before = (executable.stat().st_mtime_ns, executable.stat().st_size)
        sprite.write_bytes(png((245, 40, 110, 255)))
        first = reference('--assets-dir', assets)
        sprite.write_bytes(png((30, 220, 230, 255)))
        second = reference('--assets-dir', assets)
        assert len({REFERENCE, first, second}) == 3
        # Replace only the tree, then both; each role changes pixels independently.
        sprite.write_bytes((ROOT / 'assets/player.png').read_bytes())
        tree.write_bytes(png((90, 40, 170, 255), 18, 18))
        tree_only = reference('--assets-dir', assets)
        sprite.write_bytes(png((30, 220, 230, 255)))
        both = reference('--assets-dir', assets)
        assert len({REFERENCE, first, second, tree_only, both}) == 5
        second = both
        assert binary_before == (executable.stat().st_mtime_ns, executable.stat().st_size)

        instance = f'assets-{os.getpid()}'
        def call(*arguments):
            return json.loads(run([cli, '--format', 'json', '--project', project,
                                   '--instance', instance, *arguments], project).stdout)

        # Record with the substituted sprite, then verify in a fresh process with the same asset.
        with tempfile.TemporaryFile(mode='w+') as log:
            process = processes.Popen([str(executable), '--serve', '--project', str(project),
                '--instance', instance, '--assets-dir', str(assets), '--run-for-ms', '20000'],
                project=project, instance=instance, cwd=project, stdout=log, stderr=log)
            try:
                deadline = time.monotonic() + 5
                while not call('instances')['instances']:
                    assert process.poll() is None and time.monotonic() < deadline, 'asset session did not register'
                    time.sleep(.02)
                actions = ['right'] * 2 + ['down'] * 3 + ['right'] * 6
                for frame, action in enumerate(actions, 1):
                    call('input', str(frame), '--actions', json.dumps({action: {'kind': 'button', 'value': True}}))
                call('step', '11')
                save = call('query', 'save')['response']['value']
                recording = call('query', 'recording')['response']['value']
                assert recording['final_checksum'] == second
                record_path = project / 'custom-sprite-recording.json'
                record_path.write_text(json.dumps(recording))
                verified = json.loads(run([verifier, record_path, '--assets-dir', assets], project).stdout)
                assert verified['save'] == save and verified['checksum'] == second
                rejected = run([verifier, record_path, '--assets-dir', ROOT / 'assets'], project, success=False)
                assert 'pixels mismatch' in rejected.stderr.lower() and 'image' in rejected.stderr.lower(), rejected.stderr
                # Either mismatched source must reject fresh replay independently.
                for path in (sprite, tree):
                    retained = path.read_bytes()
                    path.write_bytes((ROOT / 'assets' / path.name).read_bytes())
                    run([verifier, record_path, '--assets-dir', assets], project, success=False)
                    path.write_bytes(retained)
                # Once loaded, disk failures cannot alter this session's art.
                sprite.write_bytes(b'broken after startup')
                tree.write_bytes(b'broken after startup')
                assert call('capture')['response']['checksum'] == second
            finally:
                try:
                    processes.graceful_shutdown(process)
                    log.seek(0)
                    assert process.returncode == 0, log.read()
                    assert not call('instances')['instances']
                finally:
                    processes.terminate(process)

        for path in (sprite, tree):
            sprite.write_bytes((ROOT / 'assets/player.png').read_bytes())
            tree.write_bytes((ROOT / 'assets/tree.png').read_bytes())
            for label, data in [('missing', None), ('corrupt', b'not a PNG'),
                                ('oversized', b'x' * (256 * 1024 + 1)),
                                ('dimensions', png((0, 0, 0, 255), 65, 1))]:
                if data is None:
                    path.unlink()
                else:
                    path.write_bytes(data)
                failed = run([executable, '--serve', '--project', project, '--instance', instance,
                              '--run-for-ms', '1000', '--assets-dir', assets], project, success=False)
                assert str(path) in failed.stderr and '--assets-dir' in failed.stderr, (label, failed.stderr)
                assert not call('instances')['instances'], label
            path.write_bytes((ROOT / 'assets' / path.name).read_bytes())
            assert reference('--assets-dir', assets) == REFERENCE

        if options.gpu:
            result = run([sys.executable, ROOT / 'scripts/build-rpg-app.py'], ROOT, phase="build")
            bundle = Path(result.stdout.strip().splitlines()[-1])
            relocated = project / 'Relocated RPG.app'
            shutil.copytree(bundle, relocated)
            bundled_player = relocated / 'Contents/MacOS/play_rpg'
            bundled_sprite = relocated / 'Contents/Resources/assets/player.png'
            assert bundled_sprite.read_bytes() == (ROOT / 'assets/player.png').read_bytes()
            for data in [(ROOT / 'assets/player.png').read_bytes(), png((100, 90, 250, 255))]:
                bundled_sprite.write_bytes(data)
                played = run([bundled_player, '--replay', '--frames', '2', '--run-for-ms', '5000'], project)
                # Exit is requested at the threshold; queued native redraws may
                # still present before the event loop finishes shutting down.
                rendered = re.search(r'rendered (\d+) GPU frames;', played.stdout)
                assert rendered and int(rendered[1]) >= 2, played.stdout
                state = json.loads(played.stdout.split('; ', 1)[1])
                assert state['collected_shards'] == 3 and state['shrine_active'] is True, state
            bundled_tree = relocated / 'Contents/Resources/assets/tree.png'
            assert bundled_tree.read_bytes() == (ROOT / 'assets/tree.png').read_bytes()
            bundled_tree.write_bytes(png((90, 40, 170, 255), 18, 18))
            run([bundled_player, '--replay', '--frames', '2', '--run-for-ms', '5000'], project)
            bundled_tree.unlink()
            failed = run([bundled_player, '--replay', '--frames', '2'], ROOT, success=False)
            assert str(bundled_tree) in failed.stderr, failed.stderr
            bundled_tree.write_bytes((ROOT / 'assets/tree.png').read_bytes())
            bundled_sprite.unlink()
            failed = run([bundled_player, '--replay', '--frames', '2'], ROOT, success=False)
            assert str(bundled_sprite) in failed.stderr, failed.stderr  # No fallback to cwd/assets.
            # An explicit external directory overrides even a missing bundled resource.
            sprite.write_bytes(png((100, 90, 250, 255)))
            run([bundled_player, '--assets-dir', assets, '--replay', '--frames', '2', '--run-for-ms', '5000'], project)
    print('RPG asset acceptance passed: exact generated/file reference, external replacement without rebuild, custom recording replay, bounded startup failures'
          + (', relocated native GPU bundle and explicit override' if options.gpu else ''))


if __name__ == '__main__':
    main()
